server/startSecureServer.js

```diff
@@ -29,27 +31,65 @@ module.exports = function (keystone, app, created, callback) {
 	if (keystone.get('ssl ca') && fs.existsSync(keystone.getPath('ssl ca'))) {
 		options.ca = fs.readFileSync(keystone.getPath('ssl ca'));
 	}
-
-	function ready (err) {
-		callback(err, message);
+	if (keystone.get('ssl pfx') && fs.existsSync(keystone.getPath('ssl pfx'))) {
+		options.pfx = fs.readFileSync(keystone.getPath('ssl pfx'));
+	}
+	if (keystone.get('ssl cert data')) {
+		options.cert = keystone.get('ssl cert');
+	}
+	if (keystone.get('ssl key data')) {
+		options.key = keystone.get('ssl key');
+	}
+	if (keystone.get('ssl ca data')) {
+		options.ca = keystone.get('ssl ca');
+	}
+	if (keystone.get('ssl pfx data')) {
+		options.pfx = keystone.get('ssl pfx');
+	}
+	if (keystone.get('ssl passphrase')) {
+		options.passphrase = keystone.get('ssl passphrase');
+	}
+	sniFunc = keystone.get('ssl sni');
+	if (sniFunc) {
+		options.SNICallback = function (host, cb) {
+			var ctx = sniFunc(host);
+			cb(null, ctx && tls.createSecureContext(ctx));
+		};
 	}
 
-	if (!options.key || !options.cert) {
-		if (ssl === 'only') {
-			console.log(keystone.get('name') + ' failed to start: invalid ssl configuration (cert and ca files required)');
-			process.exit();
+	if ((!options.key || !options.cert) && !options.pfx) {
+		if (sniFunc) {
+			// We populate the config with what sniFunc returns for localhost
+			var localCtx = sniFunc('localhost');
+			if (localCtx) {
+				for (var prop in localCtx) {
+					if (localCtx.hasOwnProperty(prop)) {
+						options[prop] = localCtx[prop];
+					}
+				}
+			}
+		}
+		if ((!options.key || !options.cert) && !options.pfx) {
```

Comment:
Yes but the options might have changed thanks to the call from sniFunc…

This is an error case, so I think code clarity is more important than not repeating the test if there is no sniFunc.